Lib/os.py

```diff
@@ -340,89 +340,97 @@ def walk(top, topdown=True, onerror=None, followlinks=False):
-                # directories if followlinks is False
-                if followlinks:
-                    walk_into = True
+                if is_dir:
+                    dirs.append(entry.name)
                 else:
-                    try:
-                        is_symlink = entry.is_symlink()
-                    except OSError:
-                        # If is_symlink() raises an OSError, consider that the
-                        # entry is not a symbolic link, same behaviour than
-                        # os.path.islink().
-                        is_symlink = False
-                    walk_into = not is_symlink
-
-                if walk_into:
-                    walk_dirs.append(entry.path)
-
-    # Yield before recursion if going top down
-    if topdown:
-        yield top, dirs, nondirs
-
-        # Recurse into sub-directories
-        islink, join = path.islink, path.join
-        for dirname in dirs:
-            new_path = join(top, dirname)
-            # Issue #23605: os.path.islink() is used instead of caching
-            # entry.is_symlink() result during the loop on os.scandir() because
-            # the caller can replace the directory entry during the "yield"
-            # above.
-            if followlinks or not islink(new_path):
-                yield from _walk(new_path, topdown, onerror, followlinks)
-    else:
-        # Recurse into sub-directories
-        for new_path in walk_dirs:
-            yield from _walk(new_path, topdown, onerror, followlinks)
-        # Yield after recursion if going bottom up
-        yield top, dirs, nondirs
+                    nondirs.append(entry.name)
+
+                if not topdown and is_dir:
+                    # Bottom-up: traverse into sub-directory, but exclude symlinks to
+                    # directories if followlinks is False
+                    if followlinks:
+                        walk_into = True
+                    else:
+                        try:
+                            is_symlink = entry.is_symlink()
+                        except OSError:
+                            # If is_symlink() raises an OSError, consider that the
+                            # entry is not a symbolic link, same behaviour than
+                            # os.path.islink().
+                            is_symlink = False
+                        walk_into = not is_symlink
+
+                    if walk_into:
+                        walk_dirs.append(entry.path)
+        if cont:
+            continue
+
+        # Yield before sub-directory traversal if going top down
+        if topdown:
+            yield top, dirs, nondirs
+            # Traverse into sub-directories
+            islink, join = path.islink, path.join
+            for dirname in reversed(dirs):
+                new_path = join(top, dirname)
+                # Issue #23605: os.path.islink() is used instead of caching
```

Comment:
```suggestion
                # bpo-23605: os.path.islink() is used instead of caching
```
To prevent future confusion: #23605 is an unrelated PR, bpo-23605 is indeed about walk/scandir.